library/init/data/nat/bitwise.lean

```diff
@@ -11,15 +11,78 @@ universe u
 
 namespace nat
 
+  def bodd_div2 : ℕ → bool × ℕ
+  | 0        := (ff, 0)
+  | (succ n) :=
+    match bodd_div2 n with
+    | (ff, m) := (tt, m)
+    | (tt, m) := (ff, succ m)
+    end
+
+  def div2 (n : ℕ) : ℕ := (bodd_div2 n).2
+
+  def bodd (n : ℕ) : bool := (bodd_div2 n).1
```

Comment:
It's not that it can't reduce; it reduces but only when `unfold_lemmas` is enabled. This causes a problem for  `binary_rec_eq`, because the type of the theorem includes `f ff 0 z = z` (where the left side has type `C (bit ff 0)` and the right side has type `C 0`), so that any users of this theorem must also have `unfold_lemmas` enabled in order to reduce `bit ff 0 := 0`. If they don't, we are in the strange circumstance of having equalities embedded in constants whose type-correctness is not verifiable, and I have no idea what kind of guarantees lean can give in this situation.